### ISSUE TITLE
Race condition if downloading protoc

### ISSUE THREAD

**Issue by aran** [pikaju/dart-protoc-builder#15]
When using the feature to download protoc and the protoc dart plugin, if there are multiple .proto targets to build, `dart run build_runner build` will run with parallelism and the download process in protoc_download.dart and protoc_plugin_download.dart, as well as the precompilation process, will race. This will cause unpredictable failures due to missing or misconfigured files and `build_runner` will then cache the failures as well. Races are more likely the more .proto files there are in a project.

It's possible to improve the probability of a successful build with a hack such as filesystem-based locks, but this is still incorrect without a true system-level lock or dependency on the download processing and a larger architectural change is probably needed to make this work correctly. 

Example `protoc_plugin_download` modification:

```
import 'dart:io';

import 'package:path/path.dart' as path;

import 'utility.dart';

/// A [Directory] to store all downloaded versions of the protoc Dart plugin.
final Directory _pluginDirectory =
    Directory(path.join(temporaryDirectory.path, 'plugin'));

Uri _protocPluginUriFromVersion(String? version) {
  return Uri.parse(
      'https://github.com/google/protobuf.dart/archive/refs/tags/protoc_plugin-v$version.zip');
}

String _protoPluginName() {
  return Platform.isWindows ? 'protoc-gen-dart.bat' : 'protoc-gen-dart';
}

/// Downloads the Dart plugin for the Protobuf compiler from the GitHub Releases
/// page and extracts it to a temporary working directory.
/// Returns the path to the binaries directory that should be added to the PATH
/// environment variable for protoc to use.
Future<File> fetchProtocPlugin(
    String version, bool precompileProtocPlugin) async {
  final packages = const ['protoc_plugin', 'protobuf'];
  // Create a temporary directory for the proto plugin of the given version.
  final versionDirectory = Directory(
      path.join(_pluginDirectory.path, 'v${version.replaceAll('.', '_')}'));
  final protocPluginPackageDirectory = Directory(path.join(
    versionDirectory.path,
    'protobuf.dart-protoc_plugin-v$version',
  ));

  final protocPluginDirectory = Directory(
    path.join(
      protocPluginPackageDirectory.path,
      'protoc_plugin',
    ),
  );

  final protocPlugin = File(
    path.join(
      protocPluginDirectory.path,
      'bin',
      _protoPluginName(),
    ),
  );

  var shouldWait = await versionDirectory.exists();

  if (shouldWait) {
    for (var wait = 0; wait < 60; wait++) {
      if (!await protocPlugin.exists()) {
        await Future.delayed(Duration(seconds: 1));
      }
    }
  } else {
    // If the plugin has not been downloaded yet, download it.

    // Download and unzip the .zip file containing protoc and Google .proto files.
    await unzipUri(
      _protocPluginUriFromVersion(version),
      versionDirectory,
      // Only extract the protoc_plugin from the Protobuf Git repository.
      (file) => packages.contains(path.split(file.name)[1]),
    );

    // Fetch protoc_plugin package dependencies.
    await Future.wait(packages.map((pkg) => ProcessExtensions.runSafely(
          'dart',
          ['pub', 'get'],
          workingDirectory: path.join(protocPluginPackageDirectory.path, pkg),
        )));

    // Make plugin executable on non-Windows platforms.
    await addRunnableFlag(protocPlugin);
  }

  if (precompileProtocPlugin) {
    final precompiledName = "precompiled.exe";
    final precompiledProtocPlugin = File(
      path.join(
        protocPluginDirectory.path,
        precompiledName,
      ),
    );

    final precompileSignal = "precompile.lock";
    final precompileSignalFile = File(
      path.join(
        protocPluginDirectory.path,
        precompileSignal,
      ),
    );

    // If the signal file exists, we are creating the plugin

    if (!await precompiledProtocPlugin.exists()) {
      var shouldWait = await precompileSignalFile.exists();
      if (shouldWait) {
        for (var wait = 0;
            wait < 60 && await precompileSignalFile.exists();
            wait++) {
          await Future.delayed(Duration(seconds: 1));
        }
      } else {
        await precompileSignalFile.create();

        // Compile the entry point file.
        await ProcessExtensions.runSafely(
          'dart',
          [
            'compile',
            'exe',
            'bin/protoc_plugin.dart',
            '-o',
            precompiledName,
          ],
          workingDirectory: protocPluginDirectory.path,
        );

        // Make sure the executable is runnable on non-Windows platforms.
        await addRunnableFlag(precompiledProtocPlugin);
        await precompileSignalFile.delete();
      }
    }
    return precompiledProtocPlugin;
  } else {
    return protocPlugin;
  }
}
```

**Comment by cpfeiffer** [comment]
The fix in #18 works great. Would you mind merging it and creating a new release?

**Comment by pikaju** [comment]
It's merged and published now
